### ISSUE TITLE
PF5 Component Review

### ISSUE THREAD

**Issue by phill-85** [patternfly-yew/patternfly-yew#50]
This issue is for tracking the upgrade to PF5.

Components:
  - [x] about
  - [x] alert
  - [x] app_launcher
  - [x] avatar
  - [x] backdrop
  - [x] background - This has changed quite abit, not sure what to do here
  - [x] badge
  - [x] brand
  - [x] breadcrumb - done, but https://github.com/ctron/patternfly-yew/issues/51
  - [x] button
  - [x] card
  - [x] chip
         Done, but:
          - PFY allows icon to be set PF only allows icon for drag & drop
          - Badge component in chip cannot be set to unread. Consider supplying Badge component as Attr, or including as child elemet (As PFReact does)
  - [x] chip_group
  - [x] clipboard
  - [x] code_block
  - [x] content - Should rename to Text
  - [x] context_selector
  - [x] divider
  - [x] dl
  - [x] dropdown - there seems to be a descrepency between the html version and the react version. The react decreciated version appears to be more similar.
  - [x] empty - title markup needs looking at.
  - [x] expandable_section - pf-m-limit-width not supported.
  - [x] file_upload
  - [x] form  - in progress. Done: TextArea, TextInput, Section, select        -- Form Helper text could be refactored to use the Helper text component underneath
  - [x] helper_text
  - [x] hint
  - [x] input_group
  - [x] label
  - [x] list
  - [x] login_page
  - [x] modal
  - [x] nav
  - [x] page
      - this has been rewritten in pf5 with the old implementation deprecated. Done a quick rewrite, to use current html/css
      - pf5 react treats each variant as a separate component. 
  - [x] pagination
  - [x] panel
  - [x] popover
  - [ ] select - toggle button doesn't include pf-v5-c-button in example source, but we use the Button component which does.
  - [x] slider
  - [x] spinner
  - [x] switch
  - [x] table - issue with icon positioning when expanding
  - [x] tabs
  - [x] text - Depreciated, replace with content component
  - [x] text_input_group
  - [x] title
  - [x] toast
  - [x] toolbar
  - [x] tooltip
  - [x] tree - disabled due to incompatibility with next::Table component - e2e3d3a 

Layouts:
  - [x] bulseye
  - [x] flex  -  Flex item does not support all modifiers (in the flex modifiers) that flex does
  - [x] gallery
  - [x] grid
  - [x] split
  - [x] stack

Misc:
  - [x] Icons - Some missing. Need input on PF5 icons

AB Comparison:
  - [x] Before and after compared for all existing examples
  - [ ] New samples should be created

**Comment by phill-85** [comment]
Looking at the css for the avatar component in specific:  https://pf5.patternfly.org/components/avatar#usage , there are a few classes like `.pf-m-sm{-on-[breakpoint]}` I feel like we should implement this maybe

**Comment by phill-85** [comment]
The `next` componentes, I presume they need changing to default implementation in v5?

**Comment by ctron** [comment]
> Looking at the css for the avatar component in specific: https://pf5.patternfly.org/components/avatar#usage , there are a few classes like `.pf-m-sm{-on-[breakpoint]}` I feel like we should implement this maybe

Makes sense, but I would add new stuff in a follow up PR.

**Comment by phill-85** [comment]
> > The about dialog no longer has a div for hero section. @ctron What do you want to do about this? https://pf5.patternfly.org/components/about-modal
> 
> Looks like the are focusing on the background aspect now. Makes sense. So guess the current (overall) logic is fine, just needs to be renamed to "background style" (vs hero).

You still want the custom style, that you added back in, just applied to the main div? I'm not sure what your use case is...

**Comment by phill-85** [comment]
> > @ctron Re the `next` components. I presume they need changing to default implementation in v5?
> 
> Yea, so the `next` components will replace the current (non-next), the current will just be gone.
> 
> I didn't want to make that change on main now, as that would case some issues for you I fear. In any case, just ignore non-next stuff (which has a next variant).

I'll make the change then.

**Comment by ctron** [comment]
> > > The about dialog no longer has a div for hero section. @ctron What do you want to do about this? https://pf5.patternfly.org/components/about-modal
> > 
> > 
> > Looks like the are focusing on the background aspect now. Makes sense. So guess the current (overall) logic is fine, just needs to be renamed to "background style" (vs hero).
> 
> You still want the custom style, that you added back in, just applied to the main div? I'm not sure what your use case is...

The use case is that we have some high and low res images. Maybe there is a better way, but the most flexible one is just setting the style: https://github.com/trustification/trustification/blob/f571378360981fae6c3bb89f4d854c9a9805a44a/spog/ui/src/about.rs#L17-L20

And maybe there is a better way.

**Comment by phill-85** [comment]
@ctron Are you happy to take a look at forms, there seems to be a lot of `next` stuff in there!

**Comment by ctron** [comment]
@phill-85 Sure, I am happy to take some parts of this. Not sure how to split it up from a technical perspective, as I would like to keep the main branch in a working condition. But we will figure it out.

Also a heads-up: I will be away from keyboard for the next week and a half. So don't expect much from me during that time :wink:

**Comment by phill-85** [comment]
> @phill-85 Sure, I am happy to take some parts of this. Not sure how to split it up from a technical perspective, as I would like to keep the main branch in a working condition. But we will figure it out.
> 
> Also a heads-up: I will be away from keyboard for the next week and a half. So don't expect much from me during that time wink

I mean I can work through it, just didn't want to hold anything up, but if your'e going to be away...

**Comment by phill-85** [comment]
I'm struggling on getting the next version of the toolbar working after removing the next mod and import stuff.

**Comment by phill-85** [comment]
> @phill-85 Sure, I am happy to take some parts of this. Not sure how to split it up from a technical perspective, as I would like to keep the main branch in a working condition. But we will figure it out.
> 
> Also a heads-up: I will be away from keyboard for the next week and a half. So don't expect much from me during that time wink

I've started tackling the forms next stuff

**Comment by phill-85** [comment]
> > > > The about dialog no longer has a div for hero section. @ctron What do you want to do about this? https://pf5.patternfly.org/components/about-modal
> > > 
> > > 
> > > Looks like the are focusing on the background aspect now. Makes sense. So guess the current (overall) logic is fine, just needs to be renamed to "background style" (vs hero).
> > 
> > 
> > You still want the custom style, that you added back in, just applied to the main div? I'm not sure what your use case is...
> 
> The use case is that we have some high and low res images. Maybe there is a better way, but the most flexible one is just setting the style: https://github.com/trustification/trustification/blob/f571378360981fae6c3bb89f4d854c9a9805a44a/spog/ui/src/about.rs#L17-L20
> 
> And maybe there is a better way.

It seems that the variable --pf-c-about-modal-box__hero--lg--BackgroundImage in no longer available.....

**Comment by phill-85** [comment]
> > > > The about dialog no longer has a div for hero section. @ctron What do you want to do about this? https://pf5.patternfly.org/components/about-modal
> > > 
> > > 
> > > Looks like the are focusing on the background aspect now. Makes sense. So guess the current (overall) logic is fine, just needs to be renamed to "background style" (vs hero).
> > 
> > 
> > You still want the custom style, that you added back in, just applied to the main div? I'm not sure what your use case is...
> 
> The use case is that we have some high and low res images. Maybe there is a better way, but the most flexible one is just setting the style: https://github.com/trustification/trustification/blob/f571378360981fae6c3bb89f4d854c9a9805a44a/spog/ui/src/about.rs#L17-L20
> 
> And maybe there is a better way.

I've had a deeper look at this and the `--pf-c-about-modal-box__hero--lg--BackgroundImage` and `--pf-c-about-modal-box__hero--sm--BackgroundImage` css vars are no longer supported in PF5.

It looks like there are not any breakpoints for the background, so I think the only option is to introduce your own. I came up with this:

```
<style>
    {r#"
        @media only screen and (min-width: 576px) {
            .pf-v5-c-about-modal-box[style]  {
                --pf-v5-c-about-modal-box--BackgroundImage: url("https://www.patternfly.org/assets/images/pfbg_992.jpg") !important;
            }
        }
        @media only screen and (min-width: 992px) {
            .pf-v5-c-about-modal-box[style]  {
                --pf-v5-c-about-modal-box--BackgroundImage: url("https://www.patternfly.org/assets/images/pfbg_992@2x.jpg") !important;
            }
        }
    "#}
</style>
```

It looks like you will have to play with other styles, as PF5 only shows the background in the bottom left corner!

**Comment by ctron** [comment]
Wow, you've been busy! I will try to get up to speed and see how I can help.

**Comment by phill-85** [comment]
> Wow, you've been busy! I will try to get up to speed and see how I can help.

I got stuck on the tree component, so just commented out the mod and use statements. Got stuck in generic trait hell iirc.

**Comment by ctron** [comment]
No worries, I am super happy you already did most of the work. I owe you a few beers (or whatever you like).

I am tempted to just take that branch and create a `feature/pf_5` branch in this repo. So we can continue by adding PR to this feature branch, and the merge it into `main` once it is ready. And once all the basic stuff looks good, I would create a new `-alpha` release.

**Comment by ctron** [comment]
It looks like on task is also to add the new menu components, which replace a lot of the existing menu-style components (like AppLauncher, ContextSelector, …)

**Comment by ctron** [comment]
I fixed the icons. Looks good. I also added a way to manually extend the icon set, as there had been at least one missing in the original set.

**Comment by ctron** [comment]
Tree should be good too

**Comment by ctron** [comment]
I migrated the background too

**Comment by ctron** [comment]
I think we are good here. Thanks again @phill-85 for all the great work!
